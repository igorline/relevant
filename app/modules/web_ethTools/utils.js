import Web3 from 'web3';
import { INFURA_PROTOCOL, INFURA_NETWORK, INFURA_API_KEY } from 'core/config';

let web3;
let rpcUrl;

const defaultOptions = { _rpcUrl: getRpcUrl(), metamask: null };

export function getMetamask() {
  return typeof window !== 'undefined' ? window.ethereum : null;
}

export function getProvider(options = defaultOptions) {
  return options.metamask || options._rpcUrl !== defaultOptions._rpcUrl
    ? initProvider(options)
    : web3 || initProvider();
}

export function initProvider(options = defaultOptions) {
  const provider = createProvider(options);
  web3 = new Web3(provider);
  return web3;
}

export function getBN(value, web3Instance = web3) {
  return new web3Instance.utils.BN(value);
}

export function getRpcUrl() {
  return rpcUrl || buildRpcUrl();
}

export function createProvider(options = defaultOptions) {
  if (options.metamask) return options.metamask;
  return options._rpcUrl.slice(0, 2) === 'ws'
    ? new Web3.providers.WebsocketProvider(options._rpcUrl)
    : new Web3.providers.HttpProvider(options._rpcUrl);
}

export function buildRpcUrl(
  protocol = INFURA_PROTOCOL,
  network = INFURA_NETWORK,
  apiKey = INFURA_API_KEY
) {
  rpcUrl = `${protocol}s://${network}.infura.io/${
    protocol === 'ws' ? 'ws/' : ''
  }v3/${apiKey}`;
  return rpcUrl;
}
