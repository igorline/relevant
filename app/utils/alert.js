import { toast } from 'react-toastify';

const IS_NATIVE = process.env.WEB !== 'true';
const IS_BROWSER = process.env.BROWSER;

export const browserAlerts = {
  alert: (message, alertType) => {
    toast(message, {
      position: toast.POSITION.BOTTOM_RIGHT,
      type: alertType || 'error',
      hideProgressBar: true
    });
  }
};

export function Alert() {
  if (IS_NATIVE) return require('react-native').Alert;
  if (IS_BROWSER) return browserAlerts;

  // eslint-disable-next-line
  return { alert: (a, b) => console.log(a, ' ', b) };
}
