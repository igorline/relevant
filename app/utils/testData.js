
export const user1 = {
  _id: 111,
  image: 'userImage1.jpg',
  handle: 'handle1',
  name: 'Name1',
  relevance: { pagerank: 11 },
  balance: 111
};

export const user2 = {
  _id: 222,
  image: 'userImage2.jpg',
  handle: 'handle2',
  name: 'Name2',
  relevance: { pagerank: 22 },
  balance: 222
};


export const post1 = {
  _id: 1111,
  user: 111,
  title: 'postTitle',
  embeddedUser: user1,
  tags: ['tag11', 'tag12'],
  body: 'awesome post #1!',
  postDate: new Date(),
  data: { pagerank: 45, payout: 18 * (10 ** 18) },
  link: { image: 'link_img1.jpg', url: 'https://example.com/testPost1', domain: 'link.domain1', title: 'postTitle1' },
  type: 'post',
};

export const post2 = {
  _id: 2222,
  user: 2222,
  title: 'postTitle',
  embeddedUser: user1,
  tags: ['tag21', 'tag22'],
  body: 'awesome post #2!',
  postDate: new Date(),
  data: { pagerank: 45, payout: 18 * (10 ** 18) },
  link: { image: 'link_img2.jpg', url: 'https://example.com/testPost2', domain: 'link.domain2', title: 'postTitle2' },
  parentPost: 1111,
  type: 'comment',
};


export const usersState = {
  users: { [user1._id]: user1, [user2._id]: user2 },
  handleToId: { [user1.handle]: user1._id, [user2.handle]: user2._id }
};

export const postsState = {
  posts: { 234: post1 }
};

export const activity = {
  totlaUsers: 3,
  amount: 10,
  byUser: user2, // TODO create user prop type validation
  post: post2, // TODO create post prop type validation
  type: 'upvote',
  createdAt: new Date()
};

export const auth = {
  community: 'testCommunity',
  user: user1,
  isAuthenticated: true
};

