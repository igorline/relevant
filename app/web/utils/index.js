import * as s3 from './s3';
import * as link from './link';
import * as auth from './auth';
import * as post from './post';

export {
  s3,
  link,
  auth,
  post
};