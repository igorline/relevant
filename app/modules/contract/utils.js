import pickBy from 'lodash.pickby';
import { types } from 'core/contracts';
import { getProvider, getBN } from 'modules/web_ethTools/utils';

export function parseBN(value, web3Instance = getProvider()) {
  return value && value.get ? getBN(value.get('_hex'), web3Instance) : value;
}

// TODO -- provide better method sorting from statesauce
export const readableMethods = Object.keys(
  pickBy(
    types.methods,
    (_, method) =>
      !types.methods[method].send &&
      method !== 'balanceOf' &&
      method !== 'isMinter' &&
      method !== 'allowance' &&
      method !== 'partialSum' &&
      method !== 'nonceOf' &&
      method !== 'isOwner' &&
      method !== 'roundDecay'
  )
);
