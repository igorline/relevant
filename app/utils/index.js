import * as s3 from './s3';
import * as post from './post';
import * as api from './api';
import * as storage from './storage';
import * as numbers from './numbers';
import * as text from './text';
import * as img from './img';
import * as alert from './alert';
import * as list from './list';
import * as tween from './tween';

let nav = {}; // eslint-disable-line

if (process.env.WEB !== 'true') {
  nav = require('./nav').default;
}

export { s3, post, api, alert, list, storage, numbers, text, nav, img, tween };
