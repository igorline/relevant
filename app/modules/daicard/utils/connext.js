/* eslint no-console: 0 */
import * as Connext from 'connext';
import interval from 'interval-promise';
import * as eth from 'ethers';
import { localStorage } from 'app/utils/storage';

const tokenAbi = require('./../abi/humanToken.json');

const { Big, minBN } = Connext.big;
const { CurrencyType, CurrencyConvertable } = Connext.types;
const { getExchangeRates, hasPendingOps } = new Connext.Utils();

const hubUrlRinkeby = 'https://rinkeby.hub.connext.network/api/hub';
const hubUrlMainnet = 'https://hub.connext.network/api/hub';

// Constants for channel max/min - this is also enforced on the hub
export const DEPOSIT_ESTIMATED_GAS = Big('700000'); // 700k gas
export const HUB_EXCHANGE_CEILING = eth.constants.WeiPerEther.mul(Big(69)); // 69 TST
export const CHANNEL_DEPOSIT_MAX = eth.constants.WeiPerEther.mul(Big(30)); // 30 TST
export const MAX_GAS_PRICE = Big('20000000000'); // 20 gWei

const RPC = 'RINKEBY';

let state = {
  isInitialized: null,
  connext: null,
  browserMinimumBalance: null,
  customWeb3: null,
  tokenContract: null,
  address: null,
  hubUrl: null
};

let updateReduxState;

export async function initConnext({ updateState, reinit }) {
  try {
    if (state.isInitialized && !reinit) return state.connext;
    updateReduxState = updateState;
    setState({ isInitialized: false });

    let mnemonic = localStorage.getItem('mnemonic');
    if (!mnemonic) {
      mnemonic = eth.Wallet.createRandom().mnemonic;
      localStorage.setItem('mnemonic', mnemonic);
    }

    await setConnext({ mnemonic });
    await setTokenContract();
    addConnextHandler();
    poller();

    setState({ isInitialized: true });

    console.log(state);
    return state.connext;
  } catch (err) {
    return null;
    console.log(err); // eslint-disable-line
  }
}

function setState(newState) {
  state = {
    ...state,
    ...newState
  };
  updateReduxState(newState);
}

async function setConnext({ rpc, mnemonic }) {
  let ethprovider;
  let hubUrl;

  switch (RPC) {
    // case 'LOCALHOST':
    //   ethprovider = localProvider;
    //   hubUrl = hubUrlLocal;
    //   break;
    case 'RINKEBY':
      ethprovider = new eth.getDefaultProvider('rinkeby'); // eslint-disable-line
      hubUrl = hubUrlRinkeby;
      break;
    case 'MAINNET':
      hubUrl = hubUrlMainnet;
      ethprovider = new eth.getDefaultProvider(); // eslint-disable-line
      break;
    default:
      throw new Error(`Unrecognized rpc: ${rpc}`);
  }

  const opts = {
    hubUrl,
    mnemonic
  };

  // *** Instantiate the connext client ***
  const connext = await Connext.getConnextClient(opts);

  const address = await connext.wallet.getAddress();

  console.log('Successfully set up connext! Connext config:');
  console.log(`  - tokenAddress: ${connext.opts.tokenAddress}`);
  console.log(`  - hubAddress: ${connext.opts.hubAddress}`);
  console.log(`  - contractAddress: ${connext.opts.contractAddress}`);
  console.log(`  - ethNetworkId: ${connext.opts.ethNetworkId}`);

  return setState({
    connext,
    tokenAddress: connext.opts.tokenAddress,
    channelManagerAddress: connext.opts.contractAddress,
    hubWalletAddress: connext.opts.hubAddress,
    ethNetworkId: connext.opts.ethNetworkId,
    address,
    ethprovider
  });
}

function addConnextHandler() {
  const { connext } = state;
  let { browserMinimumBalance } = state;
  connext.on('onStateChange', async connextState => {
    try {
      console.log('Connext state changed:', connextState);
      setState({
        connextState,
        channelState: connextState.persistent.channel,
        runtime: connextState.runtime,
        exchangeRate: connextState.runtime.exchangeRate
          ? connextState.runtime.exchangeRate.rates.USD
          : 0
      });
      if (!browserMinimumBalance) {
        browserMinimumBalance = await setBrowserWalletMinimumBalance();
        setState({ browserMinimumBalance });
      }
      // checkStatus();
    } catch (err) {
      console.log(err);
    }
  });
}

async function setBrowserWalletMinimumBalance() {
  const { connextState, ethprovider } = state;
  const gasEstimateJson = await eth.utils.fetchJson({
    url: 'https://ethgasstation.info/json/ethgasAPI.json'
  });
  const providerGasPrice = await ethprovider.getGasPrice();
  // multiply gas price by two to be safe
  let currentGasPrice = Math.round((gasEstimateJson.average / 10) * 2);
  // dont let gas price be any higher than the max
  currentGasPrice = eth.utils.parseUnits(
    minBN(Big(currentGasPrice.toString()), MAX_GAS_PRICE).toString(),
    'gwei'
  );
  // unless it really needs to be: average eth gas station price w ethprovider's
  currentGasPrice = currentGasPrice.add(providerGasPrice).div(eth.constants.Two);
  console.log(`Gas Price = ${currentGasPrice}`);

  // default connext multiple is 1.5, leave 2x for safety
  const totalDepositGasWei = DEPOSIT_ESTIMATED_GAS.mul(Big(2)).mul(currentGasPrice);

  // add dai conversion
  const minConvertable = new CurrencyConvertable(
    CurrencyType.WEI,
    totalDepositGasWei,
    () => getExchangeRates(connextState)
  );
  const browserMinimumBalance = {
    wei: minConvertable.toWEI().amount,
    dai: minConvertable.toUSD().amount
  };
  setState({ browserMinimumBalance });
  return browserMinimumBalance;
}

async function setTokenContract() {
  try {
    const { tokenAddress, ethprovider } = state;
    const tokenContract = new eth.Contract(tokenAddress, tokenAbi, ethprovider);
    return setState({ tokenContract });
  } catch (e) {
    console.log('Error setting token contract');
    return console.log(e);
  }
}

async function poller() {
  await autoDeposit();
  await autoSwap();
  interval(async () => autoDeposit(), 5000);
  interval(async () => autoSwap(), 1000);
}

async function autoDeposit() {
  const {
    address,
    tokenContract,
    connextState,
    tokenAddress,
    connext,
    browserMinimumBalance,
    ethprovider
  } = state;

  if (!connext || !browserMinimumBalance) return;

  const balance = await ethprovider.getBalance(address);

  let tokenBalance = '0';
  try {
    tokenBalance = await tokenContract.balanceOf(address);
  } catch (e) {
    console.warn(
      `Error fetching token balance, are you sure the token address (addr: ${tokenAddress}) is correct for the selected network (id: ${JSON.stringify(
        await ethprovider.getNetwork()
      )}))? Error: ${e.message}`
    );
    return;
  }

  if (balance.gt('0') || tokenBalance.gt('0')) {
    const minWei = Big(browserMinimumBalance.wei);
    if (Big(balance).lt(minWei)) {
      // don't autodeposit anything under the threshold
      // update the refunding variable before returning
      // We hit this repeatedly after first deposit & we have dust left over
      // No need to clutter logs w the below
      // console.log(`Current balance is ${balance.toString()},
      // less than minBalance of ${minWei.toString()}`);
      return;
    }
    // only proceed with deposit request if you can deposit
    if (!connextState) {
      return;
    }
    if (
      // something was submitted
      connextState.runtime.deposit.submitted ||
      connextState.runtime.withdrawal.submitted ||
      connextState.runtime.collateral.submitted
    ) {
      console.log('Deposit or withdrawal transaction in progress, will not auto-deposit');
      return;
    }

    const channelDeposit = {
      amountWei: Big(balance).sub(minWei),
      amountToken: tokenBalance
    };

    if (channelDeposit.amountWei === '0' && channelDeposit.amountToken === '0') {
      return;
    }

    await state.connext.deposit({ amountWei: channelDeposit.amountWei.toString() });
  }
}

async function autoSwap() {
  const { channelState, connextState } = state;
  if (!connextState || hasPendingOps(channelState)) {
    return;
  }
  const weiBalance = Big(channelState.balanceWeiUser);
  const tokenBalance = Big(channelState.balanceTokenUser);
  if (channelState && weiBalance.gt(Big('0')) && tokenBalance.lte(HUB_EXCHANGE_CEILING)) {
    await state.connext.exchange(channelState.balanceWeiUser, 'wei');
  }
}
