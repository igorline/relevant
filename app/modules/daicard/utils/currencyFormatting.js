/* eslint-disable */
import * as Connext from 'connext';

const { CurrencyConvertable, CurrencyType, Currency } = Connext.types;
const { getExchangeRates } = new Connext.Utils();

export function getChannelBalanceInUSD(channelState, connextState, onlyTokens = true) {
  if (!connextState || !channelState) {
    return '$0.00';
  }

  const convertableTokens = new CurrencyConvertable(
    CurrencyType.BEI,
    channelState.balanceTokenUser,
    () => getExchangeRates(connextState)
  );

  if (onlyTokens) {
    return Currency.USD(convertableTokens.toUSD().amountBigNumber).format({});
  }

  const convertableWei = new CurrencyConvertable(
    CurrencyType.WEI,
    channelState.balanceWeiUser,
    () => getExchangeRates(connextState)
  );

  console.log(
    'total:',
    convertableTokens
      .toBEI()
      .amountBigNumber.plus(convertableWei.toBEI().amountBigNumber)
      .toFixed(0)
  );

  const total = new CurrencyConvertable(
    CurrencyType.BEI,
    convertableTokens
      .toBEI()
      .amountBigNumber.plus(convertableWei.toBEI().amountBigNumber),
    () => getExchangeRates(connextState)
  ).toUSD().amountBigNumber;

  return Currency.USD(total).format({});
}

export function getAmountInUSD(amount, connextState, onlyTokens = true) {
  if (!connextState || !amount) {
    return '$0.00';
  }
  const convertableTokens = new CurrencyConvertable(
    CurrencyType.BEI,
    amount.amountToken,
    () => getExchangeRates(connextState)
  );

  if (onlyTokens) {
    return Currency.USD(convertableTokens.toUSD().amountBigNumber).format({});
  }

  const convertableWei = new CurrencyConvertable(CurrencyType.WEI, amount.amountWei, () =>
    getExchangeRates(connextState)
  );

  const totalBalance = Currency.USD(
    convertableTokens.toBEI().amountBigNumber.plus(convertableWei.toBEI)
  ).format({});

  return totalBalance;
}
