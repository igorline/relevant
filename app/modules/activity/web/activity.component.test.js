import React from 'react';
import renderer from 'react-test-renderer';
import Activity from 'modules/activity/web/activity.component';
import 'jest-styled-components';
import { MemoryRouter } from 'react-router-dom';
import { auth, activity } from 'app/utils/testData';

const props = {
  singleActivity: activity,
  auth
};

test('Snapshot Post Web', () => {
  const tree = renderer.create(
    <MemoryRouter><Activity {...props} /></MemoryRouter>
  ).toJSON();
  expect(tree).toMatchSnapshot();
});
