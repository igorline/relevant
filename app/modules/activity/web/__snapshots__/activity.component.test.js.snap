// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot Activity Web 1`] = `
<div
  className="activityItem"
>
  <div>
    <div
      className="activityInner"
    >
      <div
        className="topRow"
      >
        <a
          href="/user/profile/handle2"
          onClick={[Function]}
        >
          <span
            className="avatar leftImage"
            style={
              Object {
                "backgroundImage": "url(userImage2.jpg)",
              }
            }
          />
        </a>
        <span
          className="activityText"
        >
          <a
            href="/user/profile/handle2"
            onClick={[Function]}
          >
            Name2
             
          </a>
          upvoted your comment â†’ your relevance increased by 10
        </span>
        <div
          className="stat green"
        >
          <img
            src="/img/rup.png"
          />
          <div>
            10
          </div>
        </div>
      </div>
      <div
        style={
          Object {
            "WebkitAlignItems": "stretch",
            "WebkitBoxAlign": "stretch",
            "WebkitBoxDirection": "normal",
            "WebkitBoxOrient": "horizontal",
            "WebkitFlexBasis": "auto",
            "WebkitFlexDirection": "row",
            "WebkitFlexShrink": 0,
            "alignItems": "stretch",
            "borderBottomStyle": "solid",
            "borderBottomWidth": "0px",
            "borderLeftStyle": "solid",
            "borderLeftWidth": "0px",
            "borderRightStyle": "solid",
            "borderRightWidth": "0px",
            "borderTopStyle": "solid",
            "borderTopWidth": "0px",
            "boxSizing": "border-box",
            "display": "flex",
            "flexBasis": "auto",
            "flexDirection": "row",
            "flexShrink": 0,
            "marginBottom": "0px",
            "marginLeft": "0px",
            "marginRight": "0px",
            "marginTop": "0px",
            "minHeight": "0px",
            "minWidth": "0px",
            "msFlexAlign": "stretch",
            "msFlexDirection": "row",
            "msFlexNegative": 0,
            "msFlexPreferredSize": "auto",
            "paddingBottom": "0px",
            "paddingLeft": "0px",
            "paddingRight": "0px",
            "paddingTop": "0px",
            "position": "relative",
            "zIndex": 0,
          }
        }
      >
        <a
          className=""
          target="_blank"
        >
          <div
            style={
              Object {
                "WebkitAlignItems": "stretch",
                "WebkitBoxAlign": "stretch",
                "WebkitBoxDirection": "normal",
                "WebkitBoxFlex": 1,
                "WebkitBoxOrient": "vertical",
                "WebkitFlexBasis": "0px",
                "WebkitFlexDirection": "column",
                "WebkitFlexGrow": 1,
                "WebkitFlexShrink": 1,
                "alignItems": "stretch",
                "borderBottomStyle": "solid",
                "borderBottomWidth": "0px",
                "borderLeftStyle": "solid",
                "borderLeftWidth": "0px",
                "borderRightStyle": "solid",
                "borderRightWidth": "0px",
                "borderTopStyle": "solid",
                "borderTopWidth": "0px",
                "boxSizing": "border-box",
                "display": "flex",
                "flexBasis": "0px",
                "flexDirection": "column",
                "flexGrow": 1,
                "flexShrink": 1,
                "height": "72px",
                "marginBottom": "0px",
                "marginLeft": "0px",
                "marginRight": "14.4px",
                "marginTop": "0px",
                "minHeight": "0px",
                "minWidth": "0px",
                "msFlexAlign": "stretch",
                "msFlexDirection": "column",
                "msFlexNegative": 1,
                "msFlexPositive": 1,
                "msFlexPreferredSize": "0px",
                "paddingBottom": "0px",
                "paddingLeft": "0px",
                "paddingRight": "0px",
                "paddingTop": "0px",
                "position": "relative",
                "width": "144px",
                "zIndex": 0,
              }
            }
          >
            <div
              style={
                Object {
                  "WebkitAlignItems": "stretch",
                  "WebkitBoxAlign": "stretch",
                  "WebkitBoxDirection": "normal",
                  "WebkitBoxFlex": 1,
                  "WebkitBoxOrient": "vertical",
                  "WebkitFlexBasis": "0%",
                  "WebkitFlexDirection": "column",
                  "WebkitFlexGrow": 1,
                  "WebkitFlexShrink": 1,
                  "alignItems": "stretch",
                  "backgroundImage": "linear-gradient(3.682012153860377rad,hsla(180, 100%, 50%, 1),hsla(210,      100%, 50%, 1),hsla(240, 100%, 50%, 1))",
                  "borderBottomStyle": "solid",
                  "borderBottomWidth": "0px",
                  "borderLeftStyle": "solid",
                  "borderLeftWidth": "0px",
                  "borderRightStyle": "solid",
                  "borderRightWidth": "0px",
                  "borderTopStyle": "solid",
                  "borderTopWidth": "0px",
                  "boxSizing": "border-box",
                  "display": "flex",
                  "flexBasis": "0%",
                  "flexDirection": "column",
                  "flexGrow": 1,
                  "flexShrink": 1,
                  "marginBottom": "0px",
                  "marginLeft": "0px",
                  "marginRight": "0px",
                  "marginTop": "0px",
                  "minHeight": "0px",
                  "minWidth": "0px",
                  "msFlexAlign": "stretch",
                  "msFlexDirection": "column",
                  "msFlexNegative": 1,
                  "msFlexPositive": 1,
                  "msFlexPreferredSize": "0%",
                  "paddingBottom": "0px",
                  "paddingLeft": "0px",
                  "paddingRight": "0px",
                  "paddingTop": "0px",
                  "position": "relative",
                  "zIndex": 0,
                }
              }
            />
          </div>
        </a>
        <div
          style={
            Object {
              "WebkitAlignItems": "stretch",
              "WebkitBoxAlign": "stretch",
              "WebkitBoxDirection": "normal",
              "WebkitBoxFlex": 1,
              "WebkitBoxOrient": "vertical",
              "WebkitFlexBasis": "0px",
              "WebkitFlexDirection": "column",
              "WebkitFlexGrow": 1,
              "WebkitFlexShrink": 1,
              "alignItems": "stretch",
              "borderBottomStyle": "solid",
              "borderBottomWidth": "0px",
              "borderLeftStyle": "solid",
              "borderLeftWidth": "0px",
              "borderRightStyle": "solid",
              "borderRightWidth": "0px",
              "borderTopStyle": "solid",
              "borderTopWidth": "0px",
              "boxSizing": "border-box",
              "display": "flex",
              "flexBasis": "0px",
              "flexDirection": "column",
              "flexGrow": 1,
              "flexShrink": 1,
              "marginBottom": "0px",
              "marginLeft": "0px",
              "marginRight": "0px",
              "marginTop": "0px",
              "minHeight": "0px",
              "minWidth": "0px",
              "msFlexAlign": "stretch",
              "msFlexDirection": "column",
              "msFlexNegative": 1,
              "msFlexPositive": 1,
              "msFlexPreferredSize": "0px",
              "paddingBottom": "0px",
              "paddingLeft": "0px",
              "paddingRight": "0px",
              "paddingTop": "0px",
              "position": "relative",
              "zIndex": 0,
            }
          }
        >
          <div
            dir="auto"
            flex={1}
            style={
              Object {
                "WebkitBoxFlex": 1,
                "WebkitFlexBasis": "0px",
                "WebkitFlexGrow": 1,
                "WebkitFlexShrink": 1,
                "borderBottomWidth": "0px",
                "borderLeftWidth": "0px",
                "borderRightWidth": "0px",
                "borderTopWidth": "0px",
                "boxSizing": "border-box",
                "color": "rgba(0,0,0,1.00)",
                "display": "flex",
                "flexBasis": "0px",
                "flexGrow": 1,
                "flexShrink": 1,
                "fontFamily": "HelveticaNeue-CondensedBold",
                "fontSize": "18px",
                "fontStyle": "inherit",
                "fontVariant": "inherit",
                "fontWeight": "inherit",
                "lineHeight": "24px",
                "marginBottom": "0px",
                "marginLeft": "0px",
                "marginRight": "0px",
                "marginTop": "0px",
                "msFlexNegative": 1,
                "msFlexPositive": 1,
                "msFlexPreferredSize": "0px",
                "paddingBottom": "0px",
                "paddingLeft": "0px",
                "paddingRight": "0px",
                "paddingTop": "0px",
                "position": "relative",
                "textDecoration": "none",
                "whiteSpace": "pre-wrap",
                "wordWrap": "break-word",
              }
            }
          >
            postTitle2
          </div>
          <div
            dir="auto"
            style={
              Object {
                "WebkitFlexBasis": "auto",
                "WebkitFlexShrink": 0,
                "borderBottomWidth": "0px",
                "borderLeftWidth": "0px",
                "borderRightWidth": "0px",
                "borderTopWidth": "0px",
                "boxSizing": "border-box",
                "color": "rgba(153,153,153,1.00)",
                "display": "flex",
                "flexBasis": "auto",
                "flexShrink": 0,
                "fontFamily": "HelveticaNeue",
                "fontSize": "10.8px",
                "fontStyle": "inherit",
                "fontVariant": "inherit",
                "fontWeight": "inherit",
                "lineHeight": "14.4px",
                "marginBottom": "0px",
                "marginLeft": "0px",
                "marginRight": "0px",
                "marginTop": "0px",
                "msFlexNegative": 0,
                "msFlexPreferredSize": "auto",
                "paddingBottom": "0px",
                "paddingLeft": "0px",
                "paddingRight": "0px",
                "paddingTop": "0px",
                "position": "relative",
                "textDecoration": "none",
                "whiteSpace": "pre-wrap",
                "wordWrap": "break-word",
              }
            }
          >
            <span
              dir="auto"
              style={
                Object {
                  "borderBottomWidth": "0px",
                  "borderLeftWidth": "0px",
                  "borderRightWidth": "0px",
                  "borderTopWidth": "0px",
                  "boxSizing": "border-box",
                  "color": "inherit",
                  "display": "inline",
                  "fontFamily": "inherit",
                  "fontSize": "inherit",
                  "fontStyle": "inherit",
                  "fontVariant": "inherit",
                  "fontWeight": "inherit",
                  "lineHeight": "inherit",
                  "marginBottom": "0px",
                  "marginLeft": "0px",
                  "marginRight": "0px",
                  "marginTop": "0px",
                  "paddingBottom": "0px",
                  "paddingLeft": "0px",
                  "paddingRight": "0px",
                  "paddingTop": "0px",
                  "textDecoration": "none",
                  "whiteSpace": "inherit",
                  "wordWrap": "break-word",
                }
              }
            >
              <span
                dir="auto"
                style={
                  Object {
                    "borderBottomWidth": "0px",
                    "borderLeftWidth": "0px",
                    "borderRightWidth": "0px",
                    "borderTopWidth": "0px",
                    "boxSizing": "border-box",
                    "color": "inherit",
                    "display": "inline",
                    "fontFamily": "inherit",
                    "fontSize": "inherit",
                    "fontStyle": "inherit",
                    "fontVariant": "inherit",
                    "fontWeight": "inherit",
                    "lineHeight": "inherit",
                    "marginBottom": "0px",
                    "marginLeft": "0px",
                    "marginRight": "0px",
                    "marginTop": "0px",
                    "paddingBottom": "0px",
                    "paddingLeft": "0px",
                    "paddingRight": "0px",
                    "paddingTop": "0px",
                    "textDecoration": "none",
                    "whiteSpace": "inherit",
                    "wordWrap": "break-word",
                  }
                }
              >
                <a
                  className=""
                  href="/undefined/top/tag21"
                  onClick={[Function]}
                  styles=""
                >
                  #
                  tag21
                   
                </a>
                <a
                  className=""
                  href="/undefined/top/tag22"
                  onClick={[Function]}
                  styles=""
                >
                  #
                  tag22
                   
                </a>
                <span
                  dir="auto"
                  style={
                    Object {
                      "borderBottomWidth": "0px",
                      "borderLeftWidth": "0px",
                      "borderRightWidth": "0px",
                      "borderTopWidth": "0px",
                      "boxSizing": "border-box",
                      "color": "inherit",
                      "display": "inline",
                      "fontFamily": "inherit",
                      "fontSize": "inherit",
                      "fontStyle": "inherit",
                      "fontVariant": "inherit",
                      "fontWeight": "inherit",
                      "lineHeight": "inherit",
                      "marginBottom": "0px",
                      "marginLeft": "0px",
                      "marginRight": "0px",
                      "marginTop": "0px",
                      "paddingBottom": "0px",
                      "paddingLeft": "0px",
                      "paddingRight": "0px",
                      "paddingTop": "0px",
                      "textDecoration": "none",
                      "whiteSpace": "inherit",
                      "wordWrap": "break-word",
                    }
                  }
                >
                  â€¢ 
                </span>
              </span>
            </span>
          </div>
        </div>
      </div>
    </div>
  </div>
  <div>
    <div
      className="time"
    >
      <div
        className="border"
      />
      <span>
        a few seconds ago
      </span>
    </div>
  </div>
</div>
`;
